### PR TITLE
Enabling SPARQL parameterization

### PR DESCRIPTION
**Pull request by rwalpole**

### CODE REVIEW

File: src/main/resources/sparql/get-agent-summaries.rq [hunk 1]
```diff
@@ -0,0 +1,22 @@
...
+PREFIX time: <http://www.w3.org/2006/time#>
+PREFIX todo: <http://TODO/>
+PREFIX prov: <http://www.w3.org/ns/prov#>
+
+SELECT DISTINCT ?identifier ?agentType ?currentVersion
+
+WHERE {
+    ?identifier     dct:type            ?agentType ;
+    			    ver:currentVersion  ?currentVersion .
+    ?currentVersion $p1                 ?o1 ;
```
Comment (userjd) [inline src/main/resources/sparql/get-agent-summaries.rq:19]: Should this be ?p1

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala [hunk 2]
```diff
@@ -23,7 +23,8 @@ package uk.gov.nationalarchives.omega.api.repository
 
 import org.apache.jena.query.{ Query, QueryFactory, Syntax }
```
Comment (userjd) [inline src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala:24]: unused import

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala [hunk 3]
```diff
@@ -23,7 +23,8 @@ package uk.gov.nationalarchives.omega.api.repository
...
 import uk.gov.nationalarchives.omega.api.messages.reply.LegalStatus
-import uk.gov.nationalarchives.omega.api.repository.model.AgentEntity
+import uk.gov.nationalarchives.omega.api.messages.request.ListAgentSummary
```
Comment (userjd) [inline src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala:26]: unused import

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala [hunk 4]
```diff
@@ -23,7 +23,8 @@ package uk.gov.nationalarchives.omega.api.repository
...
+import uk.gov.nationalarchives.omega.api.repository.model.{ AgentDescriptionEntity, AgentEntity, AgentSummaryEntity }
 
 import scala.io.Source
```
Comment (userjd) [inline src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala:29]: unused import

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala [hunk 5]
```diff
@@ -23,7 +23,8 @@ package uk.gov.nationalarchives.omega.api.repository
...
 import scala.util.{ Failure, Try, Using }
```
Comment (userjd) [inline src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala:30]: unused import

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala [hunk 6]
```diff
@@ -23,7 +23,8 @@ package uk.gov.nationalarchives.omega.api.repository
 
 import org.apache.jena.query.{ Query, QueryFactory, Syntax }
 import uk.gov.nationalarchives.omega.api.messages.reply.LegalStatus
-import uk.gov.nationalarchives.omega.api.repository.model.AgentEntity
+import uk.gov.nationalarchives.omega.api.messages.request.ListAgentSummary
+import uk.gov.nationalarchives.omega.api.repository.model.{ AgentDescriptionEntity, AgentEntity, AgentSummaryEntity }
```
Comment (userjd) [inline src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala:27]: unused import

File: src/main/resources/sparql/get-agent-descriptions.rq [hunk 7]
```diff
@@ -0,0 +1,55 @@
...
+        		CONCAT(?title," ",?givenName," ",?surname),
+        		CONCAT(?givenName," ",?surname),
+        		CONCAT(?title," ",?surname),
+        		CONCAT(?title," ",?givenName),
+        		?surname, ?givenName, ?title) as ?label )
+	   	}
+    }
+  	OPTIONAL {
+    	?descriptionId rdaa:P50121|rdaa:P50037 ?dateFromBNode .
+    	?dateFromBNode time:inXSDgYear ?dateFrom .
```
Comment (userjd) [inline src/main/resources/sparql/get-agent-descriptions.rq:40]: Would any mapping pattern be useful to indicate which part of the query can be parameterized? For eg. rename `?dateFrom` to use `?dateFromValue` . Similarly for boolean, URI, timestamp etc

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/OmegaRepository.scala [hunk 8]
```diff
@@ -21,19 +21,19 @@
...
 import uk.gov.nationalarchives.omega.api.messages.reply.LegalStatus
 import uk.gov.nationalarchives.omega.api.repository.model.AgentEntity
 
 import scala.util.Try
 
-class OmegaRepository(sparqlConnector: SparqlEndpointConnector) extends AbstractRepository {
+class OmegaRepository(sparqlConnector: SparqlEndpointConnector) extends AbstractRepository with RepositoryUtils {
 
   private val sparqlResourceDir = "sparql"
   private val getLegalStatusSummarySparqlResource = s"/$sparqlResourceDir/select-legal-status-summaries.rq"
```
Comment (userjd) [inline src/main/scala/uk/gov/nationalarchives/omega/api/repository/OmegaRepository.scala:35]: should this be renamed to `selectLegalStatusSummarySparqlResource`

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/OmegaRepository.scala [hunk 9]
```diff
@@ -21,19 +21,19 @@
...
-  private val getAgentSummariesSparqlResource = s"/$sparqlResourceDir/select-agent-summaries.rq"
+  private val selectAgentSummariesSparqlResource = s"/$sparqlResourceDir/select-agent-summaries.rq"
   private val getPlaceOfDepositSummariesSparqlResource = s"/$sparqlResourceDir/select-place-of-deposit-summaries.rq"
```
Comment (userjd) [inline src/main/scala/uk/gov/nationalarchives/omega/api/repository/OmegaRepository.scala:37]: should this be renamed to selectPlaceOfDepositSummariesSparqlResource

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/SparqlParams.scala [hunk 10]
```diff
@@ -0,0 +1,82 @@
...
+  dateTimes: Map[String, String] = Map.empty,
+  values: Map[String, List[Resource]] = Map.empty,
+  queryExtension: Option[String] = None
+)
+object SparqlParams extends AgentTypeMapper {
+  def from(listAgentSummary: ListAgentSummary): SparqlParams = {
+    val valuesMap = getValuesMap(listAgentSummary)
+    val uriMap = getUriMap(listAgentSummary)
+    val booleanMap = getBooleanMap(listAgentSummary)
+    val dateTimeMap = getDateTimeMap(listAgentSummary)
```
Comment (userjd) [inline src/main/scala/uk/gov/nationalarchives/omega/api/repository/SparqlParams.scala:40]: Should we add a test to parameterize `dateTime`

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/SparqlParams.scala [hunk 11]
```diff
@@ -0,0 +1,82 @@
...
+    val queryExtension = getQueryExtension(listAgentSummary)
```
Comment (userjd) [inline src/main/scala/uk/gov/nationalarchives/omega/api/repository/SparqlParams.scala:41]: Should we add a test to parameterize `queryExtension`

File: src/main/resources/sparql/get-agent-summaries.rq (hunk 1 above)
Comment (rwalpole) [reply src/main/resources/sparql/get-agent-summaries.rq]: Updated.

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala (hunk 2 above)
Comment (rwalpole) [reply src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala]: Removed.

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala (hunk 3 above)
Comment (rwalpole) [reply src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala]: Removed.

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala (hunk 4 above)
Comment (rwalpole) [reply src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala]: Removed.

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala (hunk 5 above)
Comment (rwalpole) [reply src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala]: Removed.

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala (hunk 6 above)
Comment (rwalpole) [reply src/main/scala/uk/gov/nationalarchives/omega/api/repository/AbstractRepository.scala]: Removed.

File: src/main/resources/sparql/get-agent-descriptions.rq (hunk 7 above)
Comment (rwalpole) [reply src/main/resources/sparql/get-agent-descriptions.rq:40]: I have now renamed all SPARQL variables that can currently be parameterized to include the word `param` - although `?dateFrom` isn't actually one of them at the moment.

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/OmegaRepository.scala (hunk 8 above)
Comment (rwalpole) [reply src/main/scala/uk/gov/nationalarchives/omega/api/repository/OmegaRepository.scala]: Renamed.

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/OmegaRepository.scala (hunk 9 above)
Comment (rwalpole) [reply src/main/scala/uk/gov/nationalarchives/omega/api/repository/OmegaRepository.scala]: Renamed.

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/SparqlParams.scala (hunk 10 above)
Comment (rwalpole) [reply src/main/scala/uk/gov/nationalarchives/omega/api/repository/SparqlParams.scala:40]: I have added a new test suite `SparqlParamsSpec`. Interestingly this led me to realise that `versionTimestamp` is not being correctly interpreted in the SPARQL queries currently as I missed [some of the documentation](https://github.com/nationalarchives/ctd-omega-services/blob/2a683b480bc68795a8315ed2e5c9cbc04a631b3b/schema/omega-services.async.api.yaml#L582C13-L582C13) in the schema which for some reason I can't see in AsyncAPI Studio. I have created [a new ticket](https://national-archives.atlassian.net/browse/PACT-1087) to address this.

File: src/main/scala/uk/gov/nationalarchives/omega/api/repository/SparqlParams.scala (hunk 11 above)
Comment (rwalpole) [reply src/main/scala/uk/gov/nationalarchives/omega/api/repository/SparqlParams.scala:41]: This is now being tested in `SparqlParamsSpec`.

**Pull request merged**